microsetta_admin/metadata_util.py

```diff
@@ -1,141 +1,258 @@
+
+    values = [hsi, collection_timestamp]
+    index = ['HOST_SUBJECT_ID', 'COLLECTION_TIMESTAMP']
+
+    # TODO: denote sample projects
+    observed_multiselect = set()
+    for survey in metadata['survey_answers']:
+        for shortname, answer in survey['response'].values():
+            if isinstance(answer, list):
+                for selection in answer:
+                    specific_shortname = _build_col_name(shortname, selection)
+                    values.append('true')
+                    index.append(specific_shortname)
+                    observed_multiselect.add(specific_shortname)
+            else:
+                values.append(answer)
+                index.append(shortname)
+
+    for variable, value in sample_invariants.items():
+        index.append(variable)
+        values.append(value)
+
+    return pd.Series(values, index=index, name=name), observed_multiselect
+
+
+def _fetch_barcode_metadata(sample_barcode):
+    """Query the private API to obtain per-sample metadata
+
+    Parameters
+    ----------
+    sample_barcode : str
+        The barcode to request
+
+    Returns
+    -------
+    dict
+        The survey responses associated with the sample barcode
+    dict or None
+        Any error information associated with the retreival. If an error is
+        observed, the survey responses should not be considered valid.
+    """
+    errors = None
+
+    status, response = APIRequest.get(
+        '/api/admin/metadata/samples/%s/surveys/' % sample_barcode
+    )
+    if status != 200:
+        errors = {"barcode": sample_barcode,
+                  "error": str(status) + " from api"}
+
+    return response, errors
 
 
 def _build_col_name(col_name, multiselect_answer):
-    # TODO FIXME HACK:  How to properly build column names from english answer?
-    return col_name + "_" + \
-           (multiselect_answer.replace(" ", "_").replace("/", ""))
-
-
-def find_duplicates(barcodes):
-    seen = set([])
-    dups = set([])
-    for b in barcodes:
-        if b in seen:
-            dups.add(b)
-        seen.add(b)
-    return dups
+    """For a multiselect response, form a stable metadata variable name
+
+    Parameters
+    ----------
+    col_name : str
+        The basename for the column which would correspond to the question.
+    multiselect_answer : str
+        The selected answer
+
+    Returns
+    -------
+    str
+        The formatted column name, For example, in the primary survey
+        there is a multiple select option for alcohol which includes beer
+        and wine. The basename would be "alcohol", one multiselect_answer
+        would be "beer", and the formatted column name would be
+        "alcohol_beer".
+    """
+    # replace spaces with _
+    multiselect_answer = multiselect_answer.replace(' ', '_')
+
+    # remove any non-alphanumeric character (except for _)
+    multiselect_answer = re.sub('[^0-9a-zA-Z_]+', '', multiselect_answer)
+    return f"{col_name}_{multiselect_answer}"
```

Comment:
Name mangling could theoretically result in duplicate column names in the dataframe, but I think some of the error checking for that case was removed.  I'm not familiar enough with pandas to know what will happen if there are duplicates in the index list on construction of the series, but if it crashes at that point, I'm probably okay with it.

Reply:
I think the only way taht could happen is if the object from the API contains duplicate values like `['Beer', 'Beer', 'Cider', 'Beer']`?

Reply:
Any alphanumeric characters will get stripped, so imagine a hypothetical question on blood type with answers O+, O-, AB+, AB-, A+, A-, B+, B-.  These map to only 4 column names, so we should either be able to handle it, or crash loudly, but we can't silently overwrite in the dataframe.

Reply:
Maybe worth a unit test to make sure we crash loudly?

Reply:
Fair point :)